llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp

```diff
@@ -1800,8 +1779,18 @@ OpenMPIRBuilder::InsertPointOrErrorTy OpenMPIRBuilder::createParallel(
   Instruction *PRegPreFiniTI = PRegPreFiniBB->getTerminator();
 
   InsertPointTy PreFiniIP(PRegPreFiniBB, PRegPreFiniTI->getIterator());
-  if (Error Err = FiniCB(PreFiniIP))
-    return Err;
+  if (!FiniInfo.FiniBB) {
+    if (Error Err = FiniCB(PreFiniIP))
```

Comment:
What would set `FiniInfo.FiniBB` after the callback?

If it is a call to `createBarrier` that the `FiniCB`, that's the kind of devil's contract that make the callback-driven design so bad.

Reply:
The finalisation basic block could have already been created if the body of the parallel operation contained a cancellation point or cancel. In that case we should just branch straight to the block created previously. I agree the control flow with all of the callbacks and the cancellation stack are a bit hard to follow. This is not new with this patch.

In most cases, no cancellation will have already created a finalisation block so finalisation should be generated right here as was done before this patch.

The intention here is to only run the finalisation callback once and have all exists branch to that one instance (and also to include the barrier in that unique exit so that all threads block on the same barrier).

It is not expected that the FiniCB creates FiniBB.

Reply:
Can we improve the situation a bit to make it more predictable? There seem to be two code locations where FiniCB is called and FiniBB is assigned, here or in `commonDirectiveExit`. I propose refactoring this out into a different function whose task it is to get the FiniBB: If it was already created, return it; otherwise emit one. Such as

```cpp
struct FinalizationInfo {
  ...
  BasicBlock *getOrCreateFinializationBranchTarget() {
    if (!FiniBB) {
     FiniCB(FiniIP);
     FiniBB = FinIP.getBlock();
    }
    return FiniBB;
  }
};
```